pkg/actuators/machine/instances.go

```diff
@@ -363,68 +376,142 @@ func getImageID(machine runtimeclient.ObjectKey, machineProviderConfig *alibabac
+	} else {
+		// Otherwise, the query securityGroupIDs by the tags
+		for _, sg := range machineProviderConfig.SecurityGroups {
+			if sg.ID != "" {
+				securityGroupIDs = append(securityGroupIDs, sg.ID)
+			} else {
+				if sg.Tags != nil {
+					request := ecs.CreateDescribeSecurityGroupsRequest()
+					if machineProviderConfig.VpcID != "" {
+						request.VpcId = machineProviderConfig.VpcID
+					}
+					if machineProviderConfig.ResourceGroupID != "" {
+						request.ResourceGroupId = machineProviderConfig.ResourceGroupID
+					}
+					request.RegionId = machineProviderConfig.RegionID
+					request.Tag = buildDescribeSecurityGroupsTag(sg.Tags)
+					request.Scheme = "https"
+
+					response, err := client.DescribeSecurityGroups(request)
+					if err != nil {
+						metrics.RegisterFailedInstanceCreate(&metrics.MachineLabels{
+							Name:      machine.Name,
+							Namespace: machine.Namespace,
+							Reason:    err.Error(),
+						})
+						klog.Errorf("error describing securitygroup: %v", err)
+						return nil, fmt.Errorf("error describing securitygroup: %v", err)
+					}
+
+					if len(response.SecurityGroups.SecurityGroup) < 1 {
+						klog.Errorf("no securitygroup for given tags not found")
+						return nil, fmt.Errorf("no securitygroup for given tags not found")
+					}
+
+					for _, sg := range response.SecurityGroups.SecurityGroup {
+						securityGroupIDs = append(securityGroupIDs, sg.SecurityGroupId)
+					}
+				}
+			}
+		}
+	}
 
-	request := ecs.CreateDescribeSecurityGroupsRequest()
-	request.VpcId = machineProviderConfig.VpcID
-	request.RegionId = machineProviderConfig.RegionID
-	request.SecurityGroupId = machineProviderConfig.SecurityGroupID
-	request.Scheme = "https"
+	return &securityGroupIDs, nil
+}
 
-	response, err := client.DescribeSecurityGroups(request)
-	if err != nil {
-		metrics.RegisterFailedInstanceCreate(&metrics.MachineLabels{
-			Name:      machine.Name,
-			Namespace: machine.Namespace,
-			Reason:    err.Error(),
-		})
-		klog.Errorf("error describing securitygroup: %v", err)
-		return "", fmt.Errorf("error describing securitygroup: %v", err)
+func getMaxInstancesBySecurityGroupType(securityGroupType string) int {
+	switch securityGroupType {
+	case SecurityGroupTypeNoraml:
+		return MaxInstanceOfSecurityGroupTypeNoraml
+	case SecurityGroupTypeEnterprise:
+		return MaxInstanceOfSecurityGroupTypeEnterprise
+	default:
+		return MaxInstanceOfSecurityGroupTypeNoraml
 	}
+}
+
+func buildDescribeSecurityGroupsTag(tags []alibabacloudproviderv1.Tag) *[]ecs.DescribeSecurityGroupsTag {
+	describeSecurityGroupsTag := make([]ecs.DescribeSecurityGroupsTag, len(tags))
 
-	if len(response.SecurityGroups.SecurityGroup) < 1 {
-		klog.Errorf("no securitygroup for given filters not found")
-		return "", fmt.Errorf("no securitygroup for given filters not found")
+	for _, tag := range tags {
+		describeSecurityGroupsTag = append(describeSecurityGroupsTag, ecs.DescribeSecurityGroupsTag{
```

Comment:
`describeSecurityGroupsTag[idx] = ecs.DescribeSecurityGroupsTag{`

Reply:
Let me fix it.